pkg/virt-launcher/virtwrap/cmd-server/server.go

```diff
@@ -61,221 +71,190 @@ func getErrorMessage(err error) string {
-		reply.Success = false
-		reply.Message = err.Error()
-		return nil
+	vmi, response := getVMIFromRequest(request)
+	if !response.Success {
+		return response, nil
 	}
 
-	err = s.domainManager.DeleteVMI(vmi)
-	if err != nil {
+	if err := l.domainManager.DeleteVMI(vmi); err != nil {
 		log.Log.Object(vmi).Reason(err).Errorf("Failed to signal deletion for vmi")
-		reply.Success = false
-		reply.Message = getErrorMessage(err)
-		return nil
+		response.Success = false
+		response.Message = getErrorMessage(err)
+		return response, nil
 	}
 
 	log.Log.Object(vmi).Info("Signaled vmi deletion")
-	return nil
+	return response, nil
 }
 
-func (s *Launcher) GetDomain(args *cmdclient.Args, reply *cmdclient.Reply) error {
+func (l *Launcher) GetDomain(ctx context.Context, request *cmdv1.EmptyRequest) (*cmdv1.DomainResponse, error) {
 
-	reply.Success = true
+	response := &cmdv1.DomainResponse{
+		Response: &cmdv1.Response{
+			Success: true,
+		},
+	}
 
-	list, err := s.domainManager.ListAllDomains()
+	list, err := l.domainManager.ListAllDomains()
 	if err != nil {
-		reply.Success = false
-		reply.Message = getErrorMessage(err)
-		return nil
+		response.Response.Success = false
+		response.Response.Message = getErrorMessage(err)
+		return response, nil
 	}
 
-	if len(list) == 0 {
-		reply.Domain = nil
-	} else {
-		reply.Domain = list[0]
+	if len(list) >= 0 {
+		if domain, err := json.Marshal(list[0]); err != nil {
+			log.Log.Reason(err).Errorf("Failed to marshal domain")
+			response.Response.Success = false
+			response.Response.Message = getErrorMessage(err)
+			return response, nil
+		} else {
+			response.Domain = string(domain)
+		}
 	}
 
-	return nil
+	return response, nil
 }
 
-func (s *Launcher) GetDomainStats(args *cmdclient.Args, reply *cmdclient.Reply) error {
+func (l *Launcher) GetDomainStats(ctx context.Context, request *cmdv1.EmptyRequest) (*cmdv1.DomainStatsResponse, error) {
 
-	reply.Success = true
-
-	list, err := s.domainManager.GetDomainStats()
-	if err != nil {
-		reply.Success = false
-		reply.Message = getErrorMessage(err)
-		return nil
+	response := &cmdv1.DomainStatsResponse{
+		Response: &cmdv1.Response{
+			Success: true,
+		},
 	}
 
-	if len(list) == 0 {
-		reply.DomainStats = nil
-	} else {
-		reply.DomainStats = list[0]
-	}
-
-	return nil
-}
-
-func createSocket(socketPath string) (net.Listener, error) {
-	os.RemoveAll(socketPath)
-
-	err := os.MkdirAll(filepath.Dir(socketPath), 0755)
+	list, err := l.domainManager.GetDomainStats()
 	if err != nil {
-		log.Log.Reason(err).Error("unable to create directory for unix socket")
-		return nil, err
+		response.Response.Success = false
+		response.Response.Message = getErrorMessage(err)
+		return response, nil
 	}
 
-	socket, err := net.Listen("unix", socketPath)
-
-	if err != nil {
-		log.Log.Reason(err).Error("failed to create unix socket for launcher cmd service")
-		return nil, err
+	if len(list) >= 0 {
+		if domainStats, err := json.Marshal(list[0]); err != nil {
```

Comment:
Are we sure this works? The domainStats struct doesn't have the 'json:...' attributes for marshalling/unmarshalling. 

```pkg/virt-launcher/virtwrap/stats/types.go ```

That's where the struct is defined. We'll also need a comment there to make it known that items can only be added to that struct, and not removed/modified.

Reply:
Good catch.
The json encoder uses the field names by default. With the json tags you can override this, in order to be able to change the field name without changing the json format, and to have lowercase json keys for (uppercase) public fields.
So this works out of the box, but nevertheless it makes sense to add the tags, will do.
I also see the corresponsing unit test only tests if "something" is returned, will extend it to check some actual values.